packages/validator-ajv8/src/precompiledValidator.ts

```diff
@@ -0,0 +1,162 @@
+import { ErrorObject } from 'ajv';
+import get from 'lodash/get';
+import isEqual from 'lodash/isEqual';
+import {
+  CustomValidator,
+  ErrorSchema,
+  ErrorTransformer,
+  FormContextType,
+  hashForSchema,
+  ID_KEY,
+  RJSFSchema,
+  StrictRJSFSchema,
+  toErrorList,
+  UiSchema,
+  ValidationData,
+  ValidatorType,
+} from '@rjsf/utils';
+
+import { CompiledValidateFunction, Localizer, ValidatorFunctions } from './types';
+import processRawValidationErrors, { RawValidationErrorsType } from './processRawValidationErrors';
+
+/** `ValidatorType` implementation that uses an AJV 8 precompiled validator as created by the
+ * `compileSchemaValidators()` function provided by the `@rjsf/validator-ajv8` library.
+ */
+export default class AJV8PrecompiledValidator<
+  T = any,
+  S extends StrictRJSFSchema = RJSFSchema,
+  F extends FormContextType = any
+> implements ValidatorType<T, S, F>
+{
+  /** The root schema object used to construct this validator
+   *
+   * @private
+   */
+  readonly rootSchema: S;
+
+  /** The `ValidatorFunctions` map used to construct this validator
+   *
+   * @private
+   */
+  readonly validateFns: ValidatorFunctions<T>;
+
+  /** The main validator function associated with the base schema in the `precompiledValidator`
+   *
+   * @private
+   */
+  readonly mainValidator: CompiledValidateFunction<T>;
+
+  /** The Localizer function to use for localizing Ajv errors
+   *
+   * @private
+   */
+  readonly localizer?: Localizer;
+
+  /** Constructs an `AJV8PrecompiledValidator` instance using the `validateFns` and `rootSchema`
+   *
+   * @param validateFns - The map of the validation functions that are generated by the `schemaCompile()` function
+   * @param rootSchema - The root schema that was used with the `compileSchema()` function
+   * @param [localizer] - If provided, is used to localize a list of Ajv `ErrorObject`s
+   * @throws - Error when the base schema of the precompiled validator does not have a matching validator function
+   */
+  constructor(validateFns: ValidatorFunctions<T>, rootSchema: S, localizer?: Localizer) {
+    this.rootSchema = rootSchema;
+    this.validateFns = validateFns;
+    this.localizer = localizer;
+    this.mainValidator = this.getValidator(rootSchema);
+  }
+
+  /** Returns the precompiled validator associated with the given `schema` from the map of precompiled validator
+   * functions.
+   *
+   * @param schema - The schema for which a precompiled validator function is desired
+   * @returns - The precompiled validator function associated with this schema
+   */
+  getValidator(schema: S) {
+    const key = get(schema, ID_KEY, hashForSchema(schema));
```

Comment:
We shouldn't need to compute the hash if we have an `$id`. This is not a big deal if the hash is not an expensive computation.